model/map-file/src/main/java/org/keycloak/models/map/storage/file/FileMapStorage.java

```diff
@@ -149,79 +153,77 @@ public Crud(Class<V> entityClass, Function<String, Path> dataDirectoryFunc, Func
               .ifPresent(key -> this.fieldPredicates.replace(key, (builder, op, params) -> builder));
         }
 
-        protected Path getPathForSanitizedId(Path sanitizedIdPath) {
-            final Path dataDirectory = getDataDirectory();
-            final Path dataDirectoryWithChildren = dataDirectory.resolve(sanitizedIdPath).getParent();
-
-            if (! Files.isDirectory(dataDirectoryWithChildren)) {
-                try {
-                    Files.createDirectories(dataDirectoryWithChildren);
-                } catch (IOException ex) {
-                    throw new IllegalStateException("Directory does not exist and cannot be created: " + dataDirectory, ex);
+        protected Path getPathForEscapedId(String[] escapedIdPathArray) {
+            Path parentDirectory = getDataDirectory();
+            Path targetPath = parentDirectory;
+            for (String path : escapedIdPathArray) {
+                targetPath = targetPath.resolve(path).normalize();
+                if (! targetPath.getParent().equals(parentDirectory)) {
+                    LOG.warnf("Path traversal detected: %s", Arrays.toString(escapedIdPathArray));
+                    return null;
                 }
+                parentDirectory = targetPath;
             }
-            return dataDirectoryWithChildren.resolve(sanitizedIdPath.getFileName() + FILE_SUFFIX);
+
+            return targetPath.resolveSibling(targetPath.getFileName() + FILE_SUFFIX);
         }
 
-        protected Path getPathForSanitizedId(String sanitizedId) {
-            if (sanitizedId == null) {
-                throw new IllegalStateException("Invalid ID to sanitize");
+        protected Path getPathForEscapedId(String escapedId) {
+            if (escapedId == null) {
+                throw new IllegalStateException("Invalid ID to escape");
             }
 
-            return getPathForSanitizedId(Path.of(sanitizedId));
+            String[] escapedIdArray = ID_COMPONENT_SEPARATOR_PATTERN.split(escapedId);
+            return getPathForEscapedId(escapedIdArray);
         }
 
-        protected String sanitizeId(String id) {
-            Objects.requireNonNull(id, "ID must be non-null");
+        // Percent sign + Unix (/) and https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file reserved characters
+        private static final Pattern RESERVED_CHARACTERS = Pattern.compile("[%<:>\"/\\\\|?*=]");
+        private static final String ID_COMPONENT_SEPARATOR = ":";
```

Comment:
This is needed for React admin console to work as it does not handle slash and percent characters. See keycloak/keycloak-ui#4288.